src/dac.rs

```diff
@@ -0,0 +1,179 @@
+    TwelveL,
+    /// 12-bit precision, right-aligned.
+    TwelveR,
+}
+
+#[derive(Clone, Copy)]
+/// Select an external event to trigger the DAC.
+pub enum Trigger {
+    Timer2,
+    Timer3Or8,
+    Timer4,
+    Timer6,
+    Timer7,
+    Timer15,
+    Exti9,
+    SoftwareTrigger,
+}
+
+impl Trigger {
+    fn bits(&self) -> u8 {
+        match self {
+            Self::Timer6 => 0b000,
+            Self::Timer3Or8 => 0b001,
+            Self::Timer7 => 0b010,
+            Self::Timer15 => 0b011,
+            Self::Timer2 => 0b100,
+            Self::Timer4 => 0b101,
+            Self::Exti9 => 0b110,
+            Self::SoftwareTrigger => 0b111,
+        }
+    }
+}
+
+pub struct Dac {
+    regs: pac::DAC,
+    channel: Channel,
+    bits: DacBits,
+    vref: f32,
+}
+
+impl Dac {
+    /// Create a new DAC instance
+    pub fn new<P: Pin>(regs: pac::DAC, _pin: P, bits: DacBits, vref: f32) -> Self {
+        Self {
+            regs,
+            channel: P::CHANNEL,
+            bits,
+            vref,
+        }
+    }
+
+    /// Enable the DAC.
+    pub fn enable(&mut self, apb1: &mut APB1) {
+        apb1.enr().modify(|_, w| w.dac1en().enabled());
+        self.regs.cr.modify(|_, w| match self.channel {
+            Channel::One => w.en1().enabled(),
+            Channel::Two => w.en2().enabled(),
+        });
+    }
+
+    /// Disable the DAC
+    pub fn disable(&mut self, apb1: &mut APB1) {
+        self.regs.cr.modify(|_, w| match self.channel {
+            Channel::One => w.en1().disabled(),
+            Channel::Two => w.en2().disabled(),
+        });
+        apb1.enr().modify(|_, w| w.dac1en().disabled());
+    }
+
+    /// Set the DAC value as an integer.
+    pub fn set_value(&mut self, val: u32) {
+        match self.channel {
+            Channel::One => match self.bits {
+                DacBits::EightR => self.regs.dhr8r1.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveL => self.regs.dhr12l1.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveR => self.regs.dhr12r1.modify(|_, w| unsafe { w.bits(val) }),
+            },
+            Channel::Two => match self.bits {
+                DacBits::EightR => self.regs.dhr8r2.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveL => self.regs.dhr12l2.modify(|_, w| unsafe { w.bits(val) }),
+                DacBits::TwelveR => self.regs.dhr12r2.modify(|_, w| unsafe { w.bits(val) }),
+            },
+        }
+    }
+
+    /// Set the DAC voltage.
+    pub fn set_voltage(&mut self, volts: f32) {
+        let val = match self.bits {
+            DacBits::EightR => ((volts / self.vref) * 255.) as u32,
+            DacBits::TwelveL => ((volts / self.vref) * 4_095.) as u32,
+            DacBits::TwelveR => ((volts / self.vref) * 4_095.) as u32,
+        };
+
+        self.set_value(val);
+    }
+
+    // Select and activate a trigger. See f303 Reference manual, section 16.5.4.
+    pub fn set_trigger(&mut self, trigger: Trigger) {
```

Comment:
I wonder if we need a "`get_set_trigger`" function to check which trigger is set?
Not really needed, more like a nice to have.

Reply:
Unless we have a convincing use-case for it now, I would leave it out for simplicity. We can always add it later if it's actually needed.